### PR TITLE
feat: add new `enforce` option to `lines-between-class-members`

### PR DESCRIPTION
**Pull request by snitin315**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)
Fixes https://github.com/eslint/eslint/issues/17082

The `lines-between-class-members` can now expect an object with a property named `enforce` as the first option.

* `Object`: An object with a property named `enforce`. The enforce property should be an array of objects, each specifying the configuration for enforcing empty lines between specific pairs of class members.
    * **enforce**: You can supply any number of configurations. If a statement pair matches multiple configurations, the last matched configuration will be used. Each object should have the following properties:
        * **blankLine**: Can be set to either `"always"` or `"never"`, indicating whether a blank line should be enforced between the specified members.
        * **prev**: Specifies the type of the preceding class member. It can be `"method"` for class methods, `"field"` for class fields, or `"*"` for any class member.
        * **next**: Specifies the type of the following class member. It follows the same options as `prev`.

```js
// disallows blank lines between methods
/*eslint lines-between-class-members: [
    "error",
    {
      enforce: [
        { blankLine: "never", prev: "method", next: "method" }
      ]
    },
]*/

class MyClass {
  constructor(height, width) {
      this.height = height;
      this.width = width;
  }

  fieldA = 'Field A';

  #fieldB = 'Field B';

  method1() {}

  get area() { // error
    return this.method1();
  }

  method2() {} // error
}
```
#### Is there anything you'd like reviewers to focus on?

Should we merge both the object options into one and later in v9 remove the old syntax altogether?

### CODE REVIEW

**Comment by snitin315** [comment]
@nzakas Yes https://github.com/eslint/eslint/issues/17082, I forgot to link the issue in the description. I just need to add some docs and it will be ready for review.

File: lib/rules/lines-between-class-members.js [hunk 1]
```diff
@@ -29,7 +44,30 @@ module.exports = {
...
+                                        blankLine: { enum: ["always", "never"] },
+                                        prev: { enum: ["method", "field", "*"] },
+                                        next: { enum: ["method", "field", "*"] }
+                                    },
+                                    additionalProperties: false,
+                                    required: ["blankLine", "prev", "next"]
+                                }
+                            }
+                        },
+                        additionalProperties: false
```
Comment (ota-meshi) [inline lib/rules/lines-between-class-members.js:65]: I think we should make the enforce property a required property and require at least one element.

```suggestion
                        type: "object",
                        properties: {
                            enforce: {
                                type: "array",
                                items: {
                                    type: "object",
                                    properties: {
                                        blankLine: { enum: ["always", "never"] },
                                        prev: { enum: ["method", "field", "*"] },
                                        next: { enum: ["method", "field", "*"] }
                                    },
                                    additionalProperties: false,
                                    required: ["blankLine", "prev", "next"]
                                },
                                minItems: 1
                            }
                        },
                        additionalProperties: false,
                        required: ["enforce"]
```

File: lib/rules/lines-between-class-members.js [hunk 2]
```diff
@@ -55,6 +93,7 @@ module.exports = {
         options[0] = context.options[0] || "always";
         options[1] = context.options[1] || { exceptAfterSingleLine: false };
 
+        const configureList = typeof options[0] === "object" && options[0].enforce ? options[0].enforce : [];
```
Comment (ota-meshi) [inline lib/rules/lines-between-class-members.js:96]: I think that if `options[0]` is a string, assigning `[{ blankLine: options[0], prev: "*", next: "*" }]`, can remove the L261 statement.
What do you think?

**Review by ota-meshi — CHANGES_REQUESTED** [review]

File: lib/rules/lines-between-class-members.js (hunk 1 above)
Comment (snitin315) [reply lib/rules/lines-between-class-members.js:65]: Makes sense.

File: lib/rules/lines-between-class-members.js (hunk 2 above)
Comment (snitin315) [reply lib/rules/lines-between-class-members.js]: @ota-meshi yes, I've refactored the code.

File: lib/rules/lines-between-class-members.js (hunk 1 above)
Comment (snitin315) [reply lib/rules/lines-between-class-members.js]: Updated

File: lib/rules/lines-between-class-members.js [hunk 3]
```diff
@@ -159,20 +232,34 @@ module.exports = {
...
 
-                    if ((options[0] === "always" && !skip && !isPadded) ||
-                        (options[0] === "never" && isPadded)) {
-                        context.report({
-                            node: body[i + 1],
-                            messageId: isPadded ? "never" : "always",
-                            fix(fixer) {
-                                if (hasTokenInPadding) {
-                                    return null;
+                    if (configureList.length > 0) {
```
Comment (mdjermanovic) [inline lib/rules/lines-between-class-members.js:235]: This condition seems redundant since `configureList` is always a non-empty array?

Comment (snitin315) [reply lib/rules/lines-between-class-members.js:235]: Yes, removed.

**Review by ota-meshi — APPROVED** [review]
LGTM! Thank you!

**Review by mdjermanovic — COMMENTED** [review]
Few small suggestions, then LGTM.

File: lib/rules/lines-between-class-members.js [hunk 4]
```diff
@@ -144,6 +185,38 @@ module.exports = {
...
+
+        /**
+         * Finds the last matched configuration from the configureList.
+         * @param {ASTNode} prevNode The previous statement to match.
+         * @param {ASTNode} nextNode The current statement to match.
+         * @returns {Object} The tester of the last matched configure.
+         * @private
+         */
+        function getPaddingType(prevNode, nextNode) {
+            for (let i = configureList.length - 1; i >= 0; --i) {
```
[the remark this answers is not in the document]
Comment (snitin315) [reply lib/rules/lines-between-class-members.js:207]: Added in https://github.com/eslint/eslint/pull/17462/commits/ea4fb7fbd458d3d83f91c18ee1259b2947ead0f5

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**